src/main/java/com/example/petbuddybackend/service/care/state/CareStateMachine.java

```diff
@@ -0,0 +1,99 @@
+package com.example.petbuddybackend.service.care.state;
+
+import com.example.petbuddybackend.entity.care.Care;
+import com.example.petbuddybackend.entity.care.CareStatus;
+import com.example.petbuddybackend.entity.user.Role;
+import com.example.petbuddybackend.utils.exception.throweable.StateTransitionException;
+import org.springframework.stereotype.Service;
+
+@Service
+public final class CareStateMachine {
+
+    public static final String ACCEPTED_CARE_EDIT_MESSAGE = "Cannot edit care that has been accepted by caretaker";
+    private final TransitionManager transitionManager;
+
+    public CareStateMachine() {
+        this.transitionManager = new TransitionManager();
+        initClientTransitions(transitionManager);
+        initCaretakerTransitions(transitionManager);
+        initGlobalTransitions(transitionManager);
+    }
+
+    public Care transition(Care care, Role role, CareStatus statusToTransition) {
+        return transitionManager.transition(care, role, statusToTransition);
+    }
+
+    public void transitionToEditCare(Care care) {
+        CareStatus clientStatus = care.getClientStatus();
+        CareStatus caretakerStatus = care.getCaretakerStatus();
+
+        if(caretakerStatus.equals(CareStatus.ACCEPTED)) {
+            throw new StateTransitionException(ACCEPTED_CARE_EDIT_MESSAGE);
+        }
+
+        if(!caretakerStatus.equals(CareStatus.PENDING)) {
+            throw new StateTransitionException(new Transition(Role.CARETAKER, caretakerStatus, CareStatus.PENDING));
+        }
+
+        if(!(clientStatus.equals(CareStatus.ACCEPTED) || clientStatus.equals(CareStatus.PENDING))) {
+            throw new StateTransitionException(new Transition(Role.CLIENT, clientStatus, CareStatus.PENDING));
+        }
+
+        setBothStatuses(care, CareStatus.PENDING);
+    }
+
+    private void initClientTransitions(TransitionManager transitionManager) {
+        // Accept care
+        transitionManager.addTransition(Role.CLIENT, CareStatus.PENDING, CareStatus.ACCEPTED, this::acceptAsClient);
```

Comment:
tak naprawde, zeby klient zaakceptowal to opiekun tez musi miec accepta wczesniej

Reply:
przez to, ze przy zmianie ceny opiekunowi automatycznie wskakuje accepted

Reply:
Ok zmieni się, dzięki